### ISSUE TITLE
import statement not generated for associations with classes from other namespaces

### ISSUE THREAD

**Issue by Nava2** [umple/umple#228]
Originally reported on Google Code with ID 329
Owned by `andrew.paugh`

---
1. Use the following code
   namespace x;

namespace x.a;
class A {
}

namespace x.b;
class B {
}

association {
    1 B -> 1 A;
}
2. Generate code in Java
3. Observe that the required import x.a.\* is not added.

Umple generates correct code when association is defined within the class. The same
thing should happen when association is defined as an independent entity. By the way,
this causes some examples to generate wrong Java code such as the Airlines example.

---

Reported by `ali.fatolahi` on 2012-09-04 17:49:59

**Comment by Nava2** [comment]
Here's another example:

namespace x;

namespace x.a;
class A {
  1 -- \* B;
}

class B {
- -- 1 C;
  }

namespace x.b;
class C {
}

Umple doesn't add import x.b.\* because it only thinks of class B. But B needs a parameter
of type C

---

Reported by `ali.fatolahi` on 2012-09-04 18:47:36

**Comment by Nava2** [comment]
See also issue 171 when working on this

---

Reported by @umple on 2012-09-04 21:18:17

**Comment by Nava2** [comment]
Reported by @umple on 2012-09-27 19:38:31
- **Owner removed**
- **Labels added**: Priority-VHigh
- **Labels removed**: Priority-Critical

**Comment by Nava2** [comment]
Reported by @umple on 2012-09-28 20:37:56
- **Owner changed**: `andrew.paugh`

**Comment by Nava2** [comment]
As a UCOSP student, for this issue, I am going to attempt to document the whole process
as Tim did with issue 195, using his method as a framework.

The first thing I did is look up namespaces in the Umple User Manual, to get a better
idea of how they work.

Next, I went into UmpleOnline to test out the code. Sure enough as in the bug report,
the provided code did not result in the appropriate import being added.

The code:
  package x.a;
  import x.b.*;
is correct, for package x.a, however for package x.b we only have:
  package x.b;
with no import.

I tried the same code without the leading "namespace x;", which resulted in the same
problem. The problem seems to be in the independent association, as mentioned in the
bug report. I then learned more about associations in the user manual. I then attempted
to put the association above the class definitions, which made no difference. However,
reversing the order of the association so that it was A -> B did indeed cause the B
package to have the import, but the A package without.

Since there is no error code to look for, I'll have to look into the code and search
for where associations are analyzed.

---

Reported by `andrew.paugh` on 2012-09-29 14:07:47

**Comment by Nava2** [comment]
Reported by @umple on 2012-09-29 14:58:34
- **Labels added**: ucosp

**Comment by Nava2** [comment]
Following that, I looked at the code. But I then realized that I should write a test
first.

Using cruise.umple/test/cruise/umple/implementation/SingletonToOneTest.java as a model,
I created my own test named AssociationImportImport(.java/.ump) that creates two test
classes. The .java output is compared to the ideal output in the java/ folder, named
AssociationImportTest_Aware.java.txt and AssociationImportTest_Unaware.java.txt in
my case. With Andrew Forward's advice, I changed my class names used in the test to
Student and Mentor, since the temp files generated for those are already cleaned when
building, because they are also used in other already existing tests.

Initially, the test would fail. So I began to look at the code.

I eventually found that in UmpleInternalParser_CodeClass.ump there were two areas where
addReferencedPackage was called. The second block with myClass yourClass sure enough
had the imports reversed, with the navigable class referencing the non-navigable. Fixing
this was a simple replacement between yourClass and myClass for both if blocks.

After this, the test I wrote was able to pass, and I am ready to submit my code. I
created a patch with svn diff.

---

Reported by `andrew.paugh` on 2012-09-30 14:02:35
